compiler/src/peephole.rs

```diff
@@ -0,0 +1,182 @@
+use crate::compile::Label;
+use crate::output_stream::OutputStream;
+use arrayvec::ArrayVec;
+use rustpython_bytecode::bytecode::{self, CodeObject, Instruction, Location};
+
+const PEEPHOLE_BUFFER_SIZE: usize = 20;
+
+struct InstructionMetadata {
+    loc: Location,
+    labels: Vec<Label>,
+}
+
+impl From<Vec<InstructionMetadata>> for InstructionMetadata {
+    fn from(metas: Vec<Self>) -> Self {
+        debug_assert!(!metas.is_empty(), "`metas` must not be empty");
+        InstructionMetadata {
+            loc: metas[0].loc.clone(),
+            labels: metas
+                .into_iter()
+                .flat_map(|meta| meta.labels.into_iter())
+                .collect(),
+        }
+    }
+}
+
+pub struct PeepholeOptimizer<O: OutputStream> {
+    inner: O,
+    buffer: ArrayVec<[(Instruction, InstructionMetadata); PEEPHOLE_BUFFER_SIZE]>,
+}
+
+impl<O: OutputStream> From<CodeObject> for PeepholeOptimizer<O> {
+    fn from(code: CodeObject) -> Self {
+        Self::new(code.into())
+    }
+}
+impl<O: OutputStream> From<PeepholeOptimizer<O>> for CodeObject {
+    fn from(mut peep: PeepholeOptimizer<O>) -> Self {
+        peep.flush();
+        peep.inner.into()
+    }
+}
+
+impl<O: OutputStream> PeepholeOptimizer<O> {
+    pub fn new(inner: O) -> Self {
+        PeepholeOptimizer {
+            inner,
+            buffer: ArrayVec::default(),
+        }
+    }
+
+    fn inner_emit(inner: &mut O, instruction: Instruction, meta: InstructionMetadata) {
+        inner.emit(instruction, meta.loc);
+        for label in meta.labels {
+            inner.set_label(label);
+        }
+    }
+
+    fn emit(&mut self, instruction: Instruction, meta: InstructionMetadata) {
+        if self.buffer.is_full() {
+            let (instr, meta) = self.buffer.remove(0);
+            Self::inner_emit(&mut self.inner, instr, meta);
+        }
+        // safe because we just checked that: if full then remove one element from it
+        unsafe { self.buffer.push_unchecked((instruction, meta)) };
```

Comment:
Is this a performance improvement? I would remove the unsafe block and take the hit of a couple extra instructions. Note that we might do a search in all files for unsafe code, and this would show up. I think it should be a good goal to have no unsafe code, or at least as little as possible.